core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForSerializationBuildStep.java

```diff
@@ -0,0 +1,216 @@
+import org.jboss.jandex.Type.Kind;
+import org.jboss.logging.Logger;
+
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
+import io.quarkus.deployment.util.JandexUtil;
+import io.quarkus.runtime.annotations.RegisterForSerialization;
+
+public class RegisterForSerializationBuildStep {
+
+    private static final Logger log = Logger.getLogger(RegisterForSerializationBuildStep.class);
+
+    @Inject
+    CombinedIndexBuildItem combinedIndexBuildItem;
+
+    @BuildStep
+    public void build(BuildProducer<ReflectiveHierarchyBuildItem> reflectiveClassHierarchy) {
+        for (AnnotationInstance annotation : combinedIndexBuildItem.getIndex()
+                .getAnnotations(DotName.createSimple(RegisterForSerialization.class.getName()))) {
+
+            boolean methods = getBooleanValue(annotation, "methods");
+            boolean fields = getBooleanValue(annotation, "fields");
+
+            AnnotationValue targetsValue = annotation.value("targets");
+            AnnotationValue classNamesValue = annotation.value("classNames");
+
+            ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+
+            Set<DotName> processedReflectiveHierarchies = new HashSet<DotName>();
+
+            ReflectiveHierarchyBuildItem.Builder builder = new ReflectiveHierarchyBuildItem.Builder();
+            builder.ignoreFieldPredicate(new IgnoreFieldPredicate(fields))
+                    .ignoreMethodPredicate(new IgnoreMethodPredicate(methods));
+
+            if (targetsValue == null && classNamesValue == null) {
+                addClassNameForReflection(reflectiveClassHierarchy, classLoader, processedReflectiveHierarchies, annotation,
+                        methods, fields, builder,
+                        annotation.target().asClass().name().toString());
+            }
+
+            if (targetsValue != null) {
+                Type[] targets = targetsValue.asClassArray();
+                for (Type type : targets) {
+                    addTypeForReflection(reflectiveClassHierarchy, classLoader, processedReflectiveHierarchies, annotation,
+                            methods, fields, builder,
+                            type);
+                }
+            }
+
+            if (classNamesValue != null) {
+                String[] classNames = classNamesValue.asStringArray();
+                for (String className : classNames) {
+                    addClassNameForReflection(reflectiveClassHierarchy, classLoader, processedReflectiveHierarchies,
+                            annotation, methods, fields,
+                            builder, className);
+                }
+            }
+        }
+    }
+
+    private void addClassNameForReflection(BuildProducer<ReflectiveHierarchyBuildItem> reflectiveClassHierarchy,
```

Comment:
Should the method be named `ForSerialization`?

Reply:
Would the RegisterForSerializationBuildStep class name be a good string to be used for the source property, as the buildSetp that is generating all this?

Or should it be the actual class name that is being processed?

Reply:
Correct, I've changed all old names now.